<h1 align="center"> Battle of the Neighbourhoods - Toronto </h1>

Author: Prudhvi Sai

This notebook contains Questions 1, 2 & 3 of the Assignment as part of course era Applied Capstone Project


In [1]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis

import json # library to handle JSON files

import requests # library to handle requests
!pip install geopy
from geopy.geocoders import Nominatim # module to convert an address into latitude and longitude values

# libraries for displaying images
from IPython.display import Image 
from IPython.core.display import HTML 
    
# tranforming json file into a pandas dataframe library
from pandas.io.json import json_normalize

! pip install folium==0.5.0
import folium # plotting library

print('Folium installed')
# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors
import matplotlib.pyplot as plt # plotting library
# backend for rendering plots within the browser
%matplotlib inline 
# import k-means from clustering stage
from sklearn.cluster import KMeans

print('Libraries imported.')

Folium installed
Libraries imported.


# Question 1

## Importing Data

In [2]:
!pip install bs4
#!pip install requests
from bs4 import BeautifulSoup # this module helps in web scrapping.
import requests  # this module helps us to download a web page
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
data  = requests.get(url).text 
soup = BeautifulSoup(data,"html5lib")  # create a soup object using the variable 'data'

Steps done

**After retreiving the URL and creating a Beautiful soup object** 

 **Firstly create a list**  

 **Later after finding the table and table data  create a dictionary called cell having 3 keys PostalCode, Borough and Neighborhood.**

**As postal code contains upto 3 characters extract that using tablerow.p.text**

 **Next use split ,strip and replace functions for getting Borough and Neighborhood information.**.

 **Append to the list**  

 **Create a dataframe with list**


In [3]:
table_contents=[]
table=soup.find('table')
for row in table.findAll('td'):
    cell = {}
    if row.span.text=='Not assigned':
        pass
    else:
        cell['PostalCode'] = row.p.text[:3]
        cell['Borough'] = (row.span.text).split('(')[0]
        cell['Neighborhood'] = (((((row.span.text).split('(')[1]).strip(')')).replace(' /',',')).replace(')',' ')).strip(' ')
        table_contents.append(cell)

# print(table_contents)
df=pd.DataFrame(table_contents)
df['Borough']=df['Borough'].replace({'Downtown TorontoStn A PO Boxes25 The Esplanade':'Downtown Toronto Stn A',
                                             'East TorontoBusiness reply mail Processing Centre969 Eastern':'East Toronto Business',
                                             'EtobicokeNorthwest':'Etobicoke Northwest','East YorkEast Toronto':'East York/East Toronto',
                                             'MississaugaCanada Post Gateway Processing Centre':'Mississauga'})

In [4]:
df.head(11)

PostalCode           Borough                      Neighborhood
0         M3A        North York                         Parkwoods
1         M4A        North York                  Victoria Village
2         M5A  Downtown Toronto         Regent Park, Harbourfront
3         M6A        North York  Lawrence Manor, Lawrence Heights
4         M7A      Queen's Park     Ontario Provincial Government
5         M9A         Etobicoke                  Islington Avenue
6         M1B       Scarborough                    Malvern, Rouge
7         M3B        North York                   Don Mills North
8         M4B         East York   Parkview Hill, Woodbine Gardens
9         M5B  Downtown Toronto          Garden District, Ryerson
10        M6B        North York                         Glencairn

In [5]:
df.shape

(103, 3)

Answer to Question 1: We have 103 rows and 3 columns

Checking for number of records where Neighbourhood is "Not assigned"

In [6]:
df.Neighborhood.str.count("Not assigned").sum()

0

# Question 2

## Lattitude and longitudes of Neighbourhood

Alternatively, as suggested in the assignment, Importing the CSV file from the URL

In [7]:
data = pd.read_csv("https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DS0701EN-SkillsNetwork/labs_v1/Geospatial_Coordinates.csv")
data

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
..          ...        ...        ...
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]

Both data.csv and our df are of same size. Before joining them through Postal Code, let's check its type

In [8]:
data.dtypes

Postal Code     object
Latitude       float64
Longitude      float64
dtype: object

In [9]:
df.dtypes

PostalCode      object
Borough         object
Neighborhood    object
dtype: object

Since both are of same type object we can join them. Make sure the postal code name matches.

In [10]:
combined_data = df.join(data.set_index('Postal Code'), on='PostalCode', how='inner')
combined_data

PostalCode                Borough  \
0          M3A             North York   
1          M4A             North York   
2          M5A       Downtown Toronto   
3          M6A             North York   
4          M7A           Queen's Park   
..         ...                    ...   
98         M8X              Etobicoke   
99         M4Y       Downtown Toronto   
100        M7Y  East Toronto Business   
101        M8Y              Etobicoke   
102        M8Z              Etobicoke   

                                          Neighborhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
3                     Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4                        Ontario Provincial Government  43.662301 -79.389494  
..                                                 ...        ...        ...  
98       The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944  
99                                Church and Wellesley  43.665860 -79.383160  
100                                     Enclave of M4L  43.662744 -79.321558  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.636258 -79.498509  
102  Mimico NW, The Queensway West, South of Bloor,...  43.628841 -79.520999  

[103 rows x 5 columns]

In [11]:
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(combined_data['Borough'].unique()),
        combined_data.shape[0]
    )
)

The dataframe has 15 boroughs and 103 neighborhoods.


#### Use geopy library to get the latitude and longitude values of Toronto City.


In order to define an instance of the geocoder, we need to define a user_agent. We will name our agent <em>ny_explorer</em>, as shown below.


In [12]:
address = 'Toronto, ON'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


Let's visualize the map of Toronto

In [13]:
neighborhoods=combined_data
# create map of Toronto using latitude and longitude values
map_newyork = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(neighborhoods['Latitude'], neighborhoods['Longitude'], neighborhoods['Borough'], neighborhoods['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_newyork)  
    
map_newyork

Next, we are going to start utilizing the Foursquare API to explore the neighborhoods and segment them.


# QUESTION 3

#### Define Foursquare Credentials and Version


In [14]:
CLIENT_ID = 'ERVIEHR4XSEAHEPZGYAMLEU2KDRVWA1JRSYQ1CMKT2DRSX2Q' # your Foursquare ID
CLIENT_SECRET = 'PHDOY5Z3D4WZXSDOR4UWXU0TWQPA2MBFVGATUCW42D14S1PU' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: ERVIEHR4XSEAHEPZGYAMLEU2KDRVWA1JRSYQ1CMKT2DRSX2Q
CLIENT_SECRET:PHDOY5Z3D4WZXSDOR4UWXU0TWQPA2MBFVGATUCW42D14S1PU


In [15]:
combined_data.rename(columns = {'Neighborhood':'Neighbourhood'}, inplace = True)
# Creating the map of Toronto
map_Toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# adding markers to map
for latitude, longitude, borough, neighbourhood in zip(combined_data['Latitude'], combined_data['Longitude'], combined_data['Borough'], combined_data['Neighbourhood']):
    label = '{}, {}'.format(neighbourhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [latitude, longitude],
        radius=5,
        popup=label,
        color='red',
        fill=True
        ).add_to(map_Toronto)  
    
map_Toronto

Initializing Foursquare API credentials

Next, we create a function to get all the venue categories in Toronto

In [16]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius
            )
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Category']
    
    return(nearby_venues)

Collecting the venues in Toronto for each Neighbourhood

In [17]:
venues_in_toronto = getNearbyVenues(combined_data['Neighbourhood'], combined_data['Latitude'], combined_data['Longitude'])

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Ontario Provincial Government
Islington Avenue
Malvern, Rouge
Don Mills North
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills South
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
The Danforth  East
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmount Park
Bayview Village
Downsview East
The Danforth

In [18]:
venues_in_toronto.shape

(1266, 5)

So we have 1317 records and 5 columns. Checking sample data


In [19]:
venues_in_toronto.head()

Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0         Parkwoods               43.753259               -79.329656   
1         Parkwoods               43.753259               -79.329656   
2         Parkwoods               43.753259               -79.329656   
3  Victoria Village               43.725882               -79.315572   
4  Victoria Village               43.725882               -79.315572   

                    Venue         Venue Category  
0                     KFC   Fast Food Restaurant  
1         Brookbanks Park                   Park  
2           Variety Store      Food & Drink Shop  
3  Victoria Village Arena           Hockey Arena  
4               Portugril  Portuguese Restaurant

Checking the Venues based on Neighbourhood

In [20]:
venues_in_toronto.groupby('Neighbourhood').head()

Neighbourhood  \
0                                             Parkwoods   
1                                             Parkwoods   
2                                             Parkwoods   
3                                      Victoria Village   
4                                      Victoria Village   
...                                                 ...   
1254  Mimico NW, The Queensway West, South of Bloor,...   
1255  Mimico NW, The Queensway West, South of Bloor,...   
1256  Mimico NW, The Queensway West, South of Bloor,...   
1257  Mimico NW, The Queensway West, South of Bloor,...   
1258  Mimico NW, The Queensway West, South of Bloor,...   

      Neighbourhood Latitude  Neighbourhood Longitude                   Venue  \
0                  43.753259               -79.329656                     KFC   
1                  43.753259               -79.329656         Brookbanks Park   
2                  43.753259               -79.329656           Variety Store   
3                  43.725882               -79.315572  Victoria Village Arena   
4                  43.725882               -79.315572               Portugril   
...                      ...                      ...                     ...   
1254               43.628841               -79.520999        South St. Burger   
1255               43.628841               -79.520999               Dollarama   
1256               43.628841               -79.520999          Healthy Planet   
1257               43.628841               -79.520999              Wingporium   
1258               43.628841               -79.520999                  Subway   

             Venue Category  
0      Fast Food Restaurant  
1                      Park  
2         Food & Drink Shop  
3              Hockey Arena  
4     Portuguese Restaurant  
...                     ...  
1254           Burger Joint  
1255         Discount Store  
1256        Supplement Shop  
1257            Wings Joint  
1258         Sandwich Place  

[428 rows x 5 columns]

So there are 405 records for each neighbourhood.

Checking for the maximum venue categories

In [21]:
venues_in_toronto.groupby('Venue Category').max()

Neighbourhood  \
Venue Category                                                        
Adult Boutique                                 Church and Wellesley   
Airport                                              Downsview East   
Airport Lounge    CN Tower, King and Spadina, Railway Lands, Har...   
Airport Service   CN Tower, King and Spadina, Railway Lands, Har...   
Airport Terminal  CN Tower, King and Spadina, Railway Lands, Har...   
...                                                             ...   
Wine Bar                                   Little Portugal, Trinity   
Wine Shop                                      Church and Wellesley   
Wings Joint       Mimico NW, The Queensway West, South of Bloor,...   
Women's Store                      Lawrence Manor, Lawrence Heights   
Yoga Studio                            The Danforth West, Riverdale   

                  Neighbourhood Latitude  Neighbourhood Longitude  \
Venue Category                                                      
Adult Boutique                 43.665860               -79.383160   
Airport                        43.737473               -79.394420   
Airport Lounge                 43.628947               -79.394420   
Airport Service                43.628947               -79.394420   
Airport Terminal               43.628947               -79.394420   
...                                  ...                      ...   
Wine Bar                       43.653206               -79.379817   
Wine Shop                      43.665860               -79.383160   
Wings Joint                    43.628841               -79.520999   
Women's Store                  43.718518               -79.453512   
Yoga Studio                    43.679557               -79.352188   

                                            Venue  
Venue Category                                     
Adult Boutique                          Seduction  
Airport           Toronto Downsview Airport (YZD)  
Airport Lounge                      Porter Lounge  
Airport Service          ORNGE - Toronto Air Base  
Airport Terminal               Porter FBO Limited  
...                                           ...  
Wine Bar                        The National Club  
Wine Shop                               Wine Rack  
Wings Joint                            Wingporium  
Women's Store                       Maximum Woman  
Yoga Studio                             YogaSpace  

[223 rows x 4 columns]

There are around 232 different types of Venue Categories. Interesting!

## One Hot encoding the venue Categories

In [22]:
toronto_venue_cat = pd.get_dummies(venues_in_toronto[['Venue Category']], prefix="", prefix_sep="")
toronto_venue_cat

Adult Boutique  Airport  Airport Lounge  Airport Service  \
0                  0        0               0                0   
1                  0        0               0                0   
2                  0        0               0                0   
3                  0        0               0                0   
4                  0        0               0                0   
...              ...      ...             ...              ...   
1261               0        0               0                0   
1262               0        0               0                0   
1263               0        0               0                0   
1264               0        0               0                0   
1265               0        0               0                0   

      Airport Terminal  American Restaurant  Aquarium  Art Gallery  \
0                    0                    0         0            0   
1                    0                    0         0            0   
2                    0                    0         0            0   
3                    0                    0         0            0   
4                    0                    0         0            0   
...                ...                  ...       ...          ...   
1261                 0                    0         0            0   
1262                 0                    0         0            0   
1263                 0                    0         0            0   
1264                 0                    0         0            0   
1265                 0                    0         0            0   

      Arts & Crafts Store  Asian Restaurant  ...  \
0                       0                 0  ...   
1                       0                 0  ...   
2                       0                 0  ...   
3                       0                 0  ...   
4                       0                 0  ...   
...                   ...               ...  ...   
1261                    0                 0  ...   
1262                    0                 0  ...   
1263                    0                 0  ...   
1264                    0                 0  ...   
1265                    0                 0  ...   

      Vegetarian / Vegan Restaurant  Video Game Store  Video Store  \
0                                 0                 0            0   
1                                 0                 0            0   
2                                 0                 0            0   
3                                 0                 0            0   
4                                 0                 0            0   
...                             ...               ...          ...   
1261                              0                 0            0   
1262                              0                 0            0   
1263                              0                 0            0   
1264                              0                 0            0   
1265                              0                 0            0   

      Vietnamese Restaurant  Warehouse Store  Wine Bar  Wine Shop  \
0                         0                0         0          0   
1                         0                0         0          0   
2                         0                0         0          0   
3                         0                0         0          0   
4                         0                0         0          0   
...                     ...              ...       ...        ...   
1261                      0                0         0          0   
1262                      0                0         0          0   
1263                      0                0         0          0   
1264                      0                0         0          0   
1265                      0                0         0          0   

      Wings Joint  Women's Store  Yoga Studio  
0               0            

Adding the neighbourhood to the encoded dataframe

In [23]:
toronto_venue_cat['Neighbourhood'] = venues_in_toronto['Neighbourhood'] 

# moving neighborhood column to the first column
fixed_columns = [toronto_venue_cat.columns[-1]] + list(toronto_venue_cat.columns[:-1])
toronto_venue_cat = toronto_venue_cat[fixed_columns]

toronto_venue_cat.head()

Neighbourhood  Adult Boutique  Airport  Airport Lounge  Airport Service  \
0         Parkwoods               0        0               0                0   
1         Parkwoods               0        0               0                0   
2         Parkwoods               0        0               0                0   
3  Victoria Village               0        0               0                0   
4  Victoria Village               0        0               0                0   

   Airport Terminal  American Restaurant  Aquarium  Art Gallery  \
0                 0                    0         0            0   
1                 0                    0         0            0   
2                 0                    0         0            0   
3                 0                    0         0            0   
4                 0                    0         0            0   

   Arts & Crafts Store  ...  Vegetarian / Vegan Restaurant  Video Game Store  \
0                    0  ...                              0                 0   
1                    0  ...                              0                 0   
2                    0  ...                              0                 0   
3                    0  ...                              0                 0   
4                    0  ...                              0                 0   

   Video Store  Vietnamese Restaurant  Warehouse Store  Wine Bar  Wine Shop  \
0            0                      0                0         0          0   
1            0                      0                0         0          0   
2            0                      0                0         0          0   
3            0                      0                0         0          0   
4            0                      0                0         0          0   

   Wings Joint  Women's Store  Yoga Studio  
0            0              0            0  
1            0              0            0  
2            0              0            0  
3            0              0            0  
4            0              0            0  

[5 rows x 224 columns]

We will group the Neighbourhoods, calculate the mean venue categories in each Neighbourhood 

In [24]:
toronto_grouped = toronto_venue_cat.groupby('Neighbourhood').mean().reset_index()
toronto_grouped.head()

Neighbourhood  Adult Boutique  Airport  \
0                                        Agincourt             0.0      0.0   
1                           Alderwood, Long Branch             0.0      0.0   
2  Bathurst Manor, Wilson Heights, Downsview North             0.0      0.0   
3                                  Bayview Village             0.0      0.0   
4                Bedford Park, Lawrence Manor East             0.0      0.0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0             0.0              0.0               0.0                 0.00   
1             0.0              0.0               0.0                 0.00   
2             0.0              0.0               0.0                 0.00   
3             0.0              0.0               0.0                 0.00   
4             0.0              0.0               0.0                 0.04   

   Aquarium  Art Gallery  Arts & Crafts Store  ...  \
0       0.0          0.0                  0.0  ...   
1       0.0          0.0                  0.0  ...   
2       0.0          0.0                  0.0  ...   
3       0.0          0.0                  0.0  ...   
4       0.0          0.0                  0.0  ...   

   Vegetarian / Vegan Restaurant  Video Game Store  Video Store  \
0                            0.0               0.0          0.0   
1                            0.0               0.0          0.0   
2                            0.0               0.0          0.0   
3                            0.0               0.0          0.0   
4                            0.0               0.0          0.0   

   Vietnamese Restaurant  Warehouse Store  Wine Bar  Wine Shop  Wings Joint  \
0                    0.0              0.0       0.0        0.0          0.0   
1                    0.0              0.0       0.0        0.0          0.0   
2                    0.0              0.0       0.0        0.0          0.0   
3                    0.0              0.0       0.0        0.0          0.0   
4                    0.0              0.0       0.0        0.0          0.0   

   Women's Store  Yoga Studio  
0            0.0          0.0  
1            0.0          0.0  
2            0.0          0.0  
3            0.0          0.0  
4            0.0          0.0  

[5 rows x 224 columns]

Let's make a function to get the top most common venue categories

In [25]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [26]:
import numpy as np

There are way too many venue categories, we can take the top 10 to cluster the neighbourhoods

In [27]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighbourhood 1st Most Common Venue  \
0                                        Agincourt        Breakfast Spot   
1                           Alderwood, Long Branch           Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North                  Bank   
3                                  Bayview Village                  Café   
4                Bedford Park, Lawrence Manor East    Italian Restaurant   

  2nd Most Common Venue      3rd Most Common Venue  \
0                Lounge  Latin American Restaurant   
1                   Pub               Dance Studio   
2           Coffee Shop                       Park   
3   Japanese Restaurant                       Bank   
4        Sandwich Place                 Restaurant   

         4th Most Common Venue       5th Most Common Venue  \
0  Eastern European Restaurant  Financial or Legal Service   
1                  Coffee Shop              Sandwich Place   
2                  Gas Station   Middle Eastern Restaurant   
3           Chinese Restaurant                 Yoga Studio   
4                  Coffee Shop     Comfort Food Restaurant   

         6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0          Filipino Restaurant                 Field  Fast Food Restaurant   
1                      Dog Run  Fast Food Restaurant        Farmers Market   
2                Shopping Mall      Sushi Restaurant        Sandwich Place   
3  Eastern European Restaurant   Filipino Restaurant                 Field   
4                Grocery Store               Butcher                  Café   

  9th Most Common Venue 10th Most Common Venue  
0        Farmers Market     Falafel Restaurant  
1    Falafel Restaurant            Event Space  
2            Restaurant     Frozen Yogurt Shop  
3  Fast Food Restaurant         Farmers Market  
4                   Pub            Pizza Place

Let's make the model to cluster our Neighbourhoods

In [28]:
# import k-means from clustering stage
from sklearn.cluster import KMeans

In [29]:
# set number of clusters
k_num_clusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=k_num_clusters, random_state=0).fit(toronto_grouped_clustering)
kmeans

KMeans(algorithm='auto', copy_x=True, init='k-means++', max_iter=300,
    n_clusters=5, n_init=10, n_jobs=None, precompute_distances='auto',
    random_state=0, tol=0.0001, verbose=0)

Checking the labelling of our model

In [30]:
kmeans.labels_[0:100]

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       4, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 4, 0, 0, 0,
       0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 2, 0, 0,
       0, 0, 0, 0, 1, 0, 0, 0, 1, 3, 0, 1], dtype=int32)

Let's add the clustering Label column to the top 10 common venue categories

In [31]:
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

Join toronto_grouped with combined_data on neighbourhood to add latitude & longitude for each neighborhood to prepare it for plotting

In [32]:
toronto_merged = combined_data

toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

toronto_merged.head()

PostalCode           Borough                     Neighbourhood   Latitude  \
0        M3A        North York                         Parkwoods  43.753259   
1        M4A        North York                  Victoria Village  43.725882   
2        M5A  Downtown Toronto         Regent Park, Harbourfront  43.654260   
3        M6A        North York  Lawrence Manor, Lawrence Heights  43.718518   
4        M7A      Queen's Park     Ontario Provincial Government  43.662301   

   Longitude  Cluster Labels 1st Most Common Venue  2nd Most Common Venue  \
0 -79.329656             0.0  Fast Food Restaurant      Food & Drink Shop   
1 -79.315572             0.0           Pizza Place  Portuguese Restaurant   
2 -79.360636             0.0           Coffee Shop                 Bakery   
3 -79.464763             0.0        Clothing Store  Vietnamese Restaurant   
4 -79.389494             0.0           Coffee Shop       Sushi Restaurant   

  3rd Most Common Venue 4th Most Common Venue       5th Most Common Venue  \
0                  Park           Yoga Studio                  Donut Shop   
1           Coffee Shop          Intersection  Financial or Legal Service   
2                  Park                   Pub                        Café   
3    Miscellaneous Shop           Coffee Shop                   Gift Shop   
4         Burrito Place           Yoga Studio              Sandwich Place   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0                 Field        Farmers Market    Falafel Restaurant   
1          Hockey Arena      Cuban Restaurant            Donut Shop   
2                   Spa        Breakfast Spot        Sandwich Place   
3              Boutique         Women's Store         Deli / Bodega   
4                   Bar          Burger Joint    Mexican Restaurant   

         9th Most Common Venue 10th Most Common Venue  
0                  Event Space   Ethiopian Restaurant  
1                        Field   Fast Food Restaurant  
2               Chocolate Shop             Restaurant  
3  Eastern European Restaurant    Filipino Restaurant  
4                         Café             Restaurant

Drop all the NaN values to prevent data skew

In [33]:
toronto_merged_nonan = toronto_merged.dropna(subset=['Cluster Labels'])

Plotting the clusters on the map

In [34]:
import matplotlib.cm as cm
import matplotlib.colors as colors

In [35]:
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(k_num_clusters)
ys = [i + x + (i*x)**2 for i in range(k_num_clusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged_nonan['Latitude'], toronto_merged_nonan['Longitude'], toronto_merged_nonan['Neighbourhood'], toronto_merged_nonan['Cluster Labels']):
    label = folium.Popup('Cluster ' + str(int(cluster) +1) + '\n' + str(poi) , parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)]
        ).add_to(map_clusters)
        
map_clusters

Let's verify each of our clusters

Cluster 1

In [36]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 0, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0               North York             0.0  Fast Food Restaurant   
1               North York             0.0           Pizza Place   
2         Downtown Toronto             0.0           Coffee Shop   
3               North York             0.0        Clothing Store   
4             Queen's Park             0.0           Coffee Shop   
..                     ...             ...                   ...   
96        Downtown Toronto             0.0           Coffee Shop   
97        Downtown Toronto             0.0                  Café   
99        Downtown Toronto             0.0   Japanese Restaurant   
100  East Toronto Business             0.0    Light Rail Station   
102              Etobicoke             0.0                Bakery   

     2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0        Food & Drink Shop                  Park           Yoga Studio   
1    Portuguese Restaurant           Coffee Shop          Intersection   
2                   Bakery                  Park                   Pub   
3    Vietnamese Restaurant    Miscellaneous Shop           Coffee Shop   
4         Sushi Restaurant         Burrito Place           Yoga Studio   
..                     ...                   ...                   ...   
96                    Café    Italian Restaurant            Restaurant   
97              Restaurant           Coffee Shop         Deli / Bodega   
99        Sushi Restaurant           Escape Room      Ramen Restaurant   
100             Comic Shop               Brewery         Garden Center   
102         Discount Store           Wings Joint         Tanning Salon   

          5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                    Donut Shop                 Field        Farmers Market   
1    Financial or Legal Service          Hockey Arena      Cuban Restaurant   
2                          Café                   Spa        Breakfast Spot   
3                     Gift Shop              Boutique         Women's Store   
4                Sandwich Place                   Bar          Burger Joint   
..                          ...                   ...                   ...   
96                  Pizza Place                   Pub                Bakery   
97                          Gym  Gym / Fitness Center                   Pub   
99                    Juice Bar            Steakhouse    Mexican Restaurant   
100                      Garden  Fast Food Restaurant        Farmers Market   
102                         Gym         Grocery Store          Burger Joint   

    8th Most Common Venue        9th Most Common Venue 10th Most Common Venue  
0      Falafel Restaurant                  Event Space   Ethiopian Restaurant  
1              Donut Shop                        Field   Fast Food Restaurant  
2          Sandwich Place               Chocolate Shop             Restaurant  
3           Deli / Bodega  Eastern European Restaurant    Filipino Restaurant  
4      Mexican Restaurant                         Café             Restaurant  
..                    ...                          ...                    ...  
96   Caribbean Restaurant               Sandwich Place     Chinese Restaurant  
97     Seafood Restaurant               Sandwich Place               Beer Bar  
99             Restaurant               Adult Boutique      Indian Restaurant  
100                  Park                  Pizza Place       Recording Studio  
102     Convenience Store                Burrito Place         Sandwich Place  

[91 rows x 12 columns]

Cluster 2

In [37]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 1, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
21              York             1.0                  Park   
52        North York             1.0                  Park   
64              York             1.0                  Park   
66        North York             1.0                  Park   
91  Downtown Toronto             1.0                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
21         Women's Store                  Pool           Yoga Studio   
52           Yoga Studio            Donut Shop   Filipino Restaurant   
64           Yoga Studio            Donut Shop   Filipino Restaurant   
66     Convenience Store            Donut Shop   Filipino Restaurant   
91            Playground                 Trail               Dog Run   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
21               Dog Run                 Field  Fast Food Restaurant   
52                 Field  Fast Food Restaurant        Farmers Market   
64                 Field  Fast Food Restaurant        Farmers Market   
66                 Field  Fast Food Restaurant        Farmers Market   
91                 Field  Fast Food Restaurant        Farmers Market   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
21        Farmers Market    Falafel Restaurant            Event Space  
52    Falafel Restaurant           Event Space   Ethiopian Restaurant  
64    Falafel Restaurant           Event Space   Ethiopian Restaurant  
66    Falafel Restaurant           Event Space   Ethiopian Restaurant  
91    Falafel Restaurant           Event Space   Ethiopian Restaurant

Cluster 3

In [38]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 2, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
98  Etobicoke             2.0                 River           Yoga Studio   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
98            Donut Shop   Filipino Restaurant                 Field   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
98  Fast Food Restaurant        Farmers Market    Falafel Restaurant   

   9th Most Common Venue 10th Most Common Venue  
98           Event Space   Ethiopian Restaurant

Cluster 4

In [39]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 3, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  2nd Most Common Venue  \
22  Scarborough             3.0           Coffee Shop  Korean BBQ Restaurant   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
22           Yoga Studio   Filipino Restaurant                 Field   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
22  Fast Food Restaurant        Farmers Market    Falafel Restaurant   

   9th Most Common Venue 10th Most Common Venue  
22           Event Space   Ethiopian Restaurant

Cluster 5

In [40]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 4, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels       1st Most Common Venue  \
57   North York             4.0              Baseball Field   
101   Etobicoke             4.0  Construction & Landscaping   

    2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
57            Yoga Studio     Fish & Chips Shop   Filipino Restaurant   
101        Baseball Field     Fish & Chips Shop   Filipino Restaurant   

    5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
57                  Field  Fast Food Restaurant        Farmers Market   
101                 Field  Fast Food Restaurant        Farmers Market   

    8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
57     Falafel Restaurant           Event Space   Ethiopian Restaurant  
101    Falafel Restaurant           Event Space   Ethiopian Restaurant

We have successfully cluster Toronto neighbourhood based on venue categories!